apps/bf_temp.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2018 BaishanCloud. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <openssl/bio.h>
+#include "apps.h"
+
+#define BIO_TYPE_TEMP_FILE      (25|BIO_TYPE_FILTER)
+static BIO_METHOD *methods_tempfile = NULL;
+
+typedef struct {
+    char *filename;
+    char *tmp_filename;
+} TEMPFILE_CTX;
+
+static int tempfile_new(BIO *b);
+static int tempfile_free(BIO *b);
+static int tempfile_read(BIO *b, char *out, int outl);
+static int tempfile_write(BIO *b, const char *in, int inl);
+static long tempfile_ctrl(BIO *b, int cmd, long num, void *ptr);
+static int tempfile_gets(BIO *b, char *buf, int size);
+static int tempfile_puts(BIO *b, const char *str);
+static void make_temp_filename(const char *orig, char *filename);
+
+BIO_METHOD *apps_bf_tempfile(void)
+{
+    if (methods_tempfile == NULL) {
+        methods_tempfile = BIO_meth_new(BIO_TYPE_TEMP_FILE,
+                                        "Temporary file filter");
+        if (methods_tempfile == NULL
+            || !BIO_meth_set_write(methods_tempfile, tempfile_write)
+            || !BIO_meth_set_read(methods_tempfile, tempfile_read)
+            || !BIO_meth_set_puts(methods_tempfile, tempfile_puts)
+            || !BIO_meth_set_gets(methods_tempfile, tempfile_gets)
+            || !BIO_meth_set_ctrl(methods_tempfile, tempfile_ctrl)
+            || !BIO_meth_set_create(methods_tempfile, tempfile_new)
+            || !BIO_meth_set_destroy(methods_tempfile, tempfile_free))
+            return NULL;
+    }
+    return methods_tempfile;
+}
+
+void apps_bf_tempfile_cleanup(void)
+{
+    BIO_meth_free(methods_tempfile);
+    methods_tempfile = NULL;
+}
+
+static int tempfile_new(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    temp = app_malloc(sizeof(*temp), "temp file");
+    temp->filename = NULL;
+    temp->tmp_filename = NULL;
+    BIO_set_data(b, temp);
+    BIO_set_init(b, 1);
+
+    return 1;
+}
+
+#ifdef _WIN32
+static int WIN32_rename(const char *from, const char *to);
+# define rename(from,to) WIN32_rename((from),(to))
+#endif
+
+static int tempfile_free(BIO *b)
+{
+    TEMPFILE_CTX *temp;
+
+    if (b == NULL)
+        return 0;
+
+    temp = BIO_get_data(b);
+    if (temp->filename != NULL && temp->tmp_filename != NULL) {
+#if !defined(__VMS)
+        rename(temp->tmp_filename, temp->filename);
+#endif
+    }
+    if (temp->filename)
+        OPENSSL_free(temp->filename);
+    if (temp->tmp_filename)
+        OPENSSL_free(temp->tmp_filename);
```

Comment:
There's no need to check if the object to be freed is non-NULL, `OPENSSL_free` does that for you.

Reply:
Hmm, it seems it's the underlying `free` that does this.